src/linear_algebra/dfinsupp.lean

```diff
@@ -189,4 +190,16 @@ lemma map_range.linear_equiv_symm (e : Π i, β₁ i ≃ₗ[R] β₂ i) :
 
 end map_range
 
+section basis
+
+/-- The direct sum of free modules is free.
+
+Note that while this is stated for `dfinsupp` not `direct_sum`, the types are defeq. -/
```

Comment:
Do you know why we have module instances for both `dfinsupp` and `direct_sum`? Should the second one become notation perhaps?

Reply:
My mental model is that `finsupp` is to `dfinsupp` as `monoid_algebra` is to `direct_sum`.

Reply:
No idea why `direct_sum` is a `def` rather than `notation`, but I was indeed slightly confused by this the first time I saw it.

Reply:
The way I understand `monoid_algebra` is as a type synonym for `finsupp` that chooses `*` to be the convolution product (rather than the pointwise product). But the direct sum of modules doesn't have a product, right?

Reply:
Mathematically they work in different categories, `monoid_algebra` is the left/right (I always forget) adjoint to the forgetful functor `algebras → monoids`, so it is like free modules (that is the adjoint to the forgetful `modules → sets`). The direct sum is the coproduct in `R-mod`, and the analogue for algebras is tensor product.

Reply:
> But the direct sum of modules doesn't have a product, right?

If those modules are submodules of the same algebra, then `direct_sum.ring` provides this product via `direct_sum.gmonoid.of_submodules`